### ISSUE TITLE
Appium burns CPU cycles while application is idle

### ISSUE THREAD

**Issue by willosser** [appium/appium#4698]
When we run our performance tests, one of the values we measure is cpu usage, using adb dumpsys cpuinfo.  When we moved our performance tests to us Appium, we noticed a much higher cpu baseline.  The cpu usage for our app when it was completely idle is show as 63%.  The same test when run with our previous test harness using uiautomator shows 0%.  To verify the problem, I did the following:

Write an Appium test that merely starts up our application and sleeps.  We received the same result

Write an Appium test that starts a small application that merely consists of two buttons that can be pressed.  The cpu usage was still at 43%.

Can we expose a capability which dials down the frequency of whatever application pinging Appium is doing?  I realize that this could possibly lead to sluggishness on the part of Appium, which is why I'm suggesting allowing the user to set the rate rather than hard-coding it.

**Comment by imurchie** [comment]
Interesting. Similar results using `adb shell top -m 10` to monitor. Adding to milestone to get dealt with.

**Comment by moizjv** [comment]
@willosser  so i looked into this the reason for this is we add UIWatchers to watch for if application crashes or not responsive to fail tests. When i commented this code i got CPU usage down from 54% to 12%. Here is the [watcher](https://github.com/appium/appium-android-bootstrap/blob/master/src/com/android/uiautomator/common/UiWatchers.java) code . We can think of adding capability of disabling this. 

Also appium should not be used to test performance. Yes we need to be careful in terms of memory leaks if any from appium and make appium more efficient.

**Comment by moizjv** [comment]
@willosser what do you think?

**Comment by willosser** [comment]
Based on my past experience writing test automation code for Android, I think this watcher code is expensive in terms of execution time.  I've written code that runs an adb command in another thread and monitors the output:
  adb -s <device_id> shell am monitor 
Any output indicates some app crashed.  In our testing, we found it might not be the app under test, so we had to parse the output to verify which app failed.

I'd rather see Appium available for performance testing, or at least as usable as possible.  The alternative is asking our developers to provide hooks in the code and output in logcat to drive such testing through adb

**Comment by bootstraponline** [comment]
I agree with providing a cap to disable the watcher code.

**Comment by jlipps** [comment]
:+1: to adding a flag to disable watchers. not sure whether it should be a cap or server flag

**Comment by bootstraponline** [comment]
The problem with flags is they are difficult to control dynamically. Maybe I want one test to have this feature off for perf measurements, and then the next test I'd like it on to make sure the crash dialogs don't mess with the results.

**Comment by willosser** [comment]
I have no need of controlling the performance dynamically.  I don't tend to mix performance tests with system tests.  On the other hand, I tried a run of our performance test with the monitoring code disabled and did not have the same stellar results as @moizjv.  My cpu usage dropped from 63% when idle to 36-44%.

**Comment by jlipps** [comment]
I'm fine making it a cap then.

**Comment by moizjv** [comment]
@willosser so commenting watcher code does bring down CPU, I can see if i can find other trouble spots. @jlipps Yes will make a cap to ignore watcher code !

**Comment by willosser** [comment]
Why don't we implement the watcher code in adb instead?  The cpu load is minimal and this would allow us to continue to monitor ANRs

**Comment by jlipps** [comment]
that's fine with me. I don't understand this whole feature well enough to make any counterargument.

**Comment by moizjv** [comment]
@willosser Yes that would be ideal, but here with the present watcher we dismiss the notification using watcher and ui automator, I am not sure how that would work using adb. We can get @bootstraponline opinion on that. For now I have added disableAndroidWathers cap that would help bring down cpu usage but with a tradeoff

**Comment by bootstraponline** [comment]
> Why don't we implement the watcher code in adb instead?

because that's impossible? There are two approaches to dealing with these types of crash dialogs. The first is to modify the android image so they don't exist (Google talked about this at GTAC). This isn't feasible for smaller entities so that's not a great option for appium. The second approach is using an automation framework, in this case uiautomator, to poll and handle the specific crash dialogs as they occur. adb simply is not able to deal with these kinds of issues.

**Comment by willosser** [comment]
"adb shell am monitor" will report the ANR.  What I haven't proven conclusively to myself yet is whether or not the ANR reported this way will also result in a screen dialog that needs to be dismissed.  Thanks for the answer.

**Comment by bootstraponline** [comment]
@willosser If you look at [the Java code](https://github.com/appium/appium/blob/9d89eea9d42d53c2d70533e8ec6bdb52e952f587/lib/devices/android/bootstrap/src/com/android/uiautomator/common/UiWatchers.java) I stand by my assessment that adb is not able to do the automatic detection and dismissal. Even if `am` is able to monitor the dialogs, I don't see a way to interact with them reliably via adb. If this is possible then I'm happy to investigate.

**Comment by willosser** [comment]
sorry @bootstraponline, my response was incomplete.  adb can detect an ANR, but not detect anything on the screen.  My idea was to detect the ANR through adb.  Once detected, search for the on-screen dialogs using uiautomator and dismiss them.  This would reduce the cpu usage while waiting for an ANR, but still detect the dialogs and dismiss them.

**Comment by bootstraponline** [comment]
I looked through the [am](https://android.googlesource.com/platform/frameworks/base/+/742a67127366c376fdf188ff99ba30b27d3bf90c/cmds/am/src/com/android/commands/am/Am.java) source code and it isn't designed for this use case. The monitor will stop upon detection and then wait for a continue/kill command. The impact would be making appium even slower than it already is today.

The am implementation is based on [IActivityController](https://android.googlesource.com/platform/frameworks/base.git/+/60d8762413e8daba5f73559786312a9ec5e3b827/core/java/android/app/IActivityController.aidl) which looks like it'd be more performant than polling. If we could use that from the bootstrap jar instead of polling, that'd be ideal.

**Comment by bootstraponline** [comment]
I ported the bits of `am` that make sense for this use case. The work is [available on this branch](https://github.com/bootstraponline/appium/commit/795b5f2ae3c1b346d5bd2a8e1c89eb131efc337f). I haven't yet integrated it with appium (MyActivityController has to be instantiated instead of the regular uiwatcher class) or tested. I think it'll work and I found out that it'll prevent the dialog from even showing up so we don't have to worry about automating the dismissal of the alert.
